samples/bluetooth/mesh/light_dimmer/src/model_handler.c

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+/**
+ * @file
+ * @brief Model handler for the Light Dimmer Controller.
+ *
+ */
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/mesh/proxy.h>
+#include <bluetooth/mesh/models.h>
+#include <dk_buttons_and_leds.h>
+#include "model_handler.h"
+#include <time.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dimmer_sample, LOG_LEVEL_INF);
+
+#define SHORT_PRESS_THRESHOLD_MS 300
+#define LONG_PRESS_TIMEOUT_SEC K_SECONDS(20)
+#define DELTA_MOVE_STEP 1000
+
+enum dim_direction { DIM_DOWN, DIM_UP };
+
+/** Context for a single Light Dimmer instance. */
+static struct dim_button {
+	/** Generic OnOff client instance for this switch */
+	struct bt_mesh_onoff_cli onoff_client;
+	/** Generic Level client instance for this switch */
+	struct bt_mesh_lvl_cli lvl_client;
+	/* Relative start time of the latest press of the button */
+	int64_t press_start_time;
+	/* Dimmer direction */
+	enum dim_direction direction;
+	/* Long press work (Dimmer mode) */
+	struct k_work_delayable long_press_start_work;
+	struct k_work_delayable long_press_stop_work;
+};
+
+/* Common handlers for dimmer buttons */
+
+static void dimmer_start(struct k_work *work)
+{
+	struct dim_button *dimmer = CONTAINER_OF(k_work_delayable_from_work(work),
+						 struct dim_button, long_press_start_work);
+
+	struct bt_mesh_lvl_move_set move_set = {
+		.delta = (dimmer->direction == DIM_UP) ? DELTA_MOVE_STEP :
+							 (int16_t)(-1 * DELTA_MOVE_STEP),
+		.transition = &(struct bt_mesh_model_transition){ .time = 500 },
+	};
+
+	LOG_INF("Start dimming %s...", (dimmer->direction == DIM_UP ? "UP" : "DOWN"));
+
+	/* As we can't know how many nodes are in a group, it doesn't
+	 * make sense to send acknowledged messages to group addresses.
+	 */
+	if (bt_mesh_model_pub_is_unicast(dimmer->lvl_client.model)) {
+		bt_mesh_lvl_cli_move_set(&dimmer->lvl_client, NULL, &move_set, NULL);
+	} else {
+		bt_mesh_lvl_cli_move_set_unack(&dimmer->lvl_client, NULL, &move_set);
+	}
+
+	k_work_reschedule(&dimmer->long_press_stop_work, LONG_PRESS_TIMEOUT_SEC);
+}
+
+static void dimmer_stop(struct k_work *work)
+{
+	struct dim_button *dimmer = CONTAINER_OF(k_work_delayable_from_work(work),
+						 struct dim_button, long_press_stop_work);
+
+	struct bt_mesh_lvl_move_set move_set = {
+		.delta = 0,
+		.transition = NULL,
+	};
+
+	if (bt_mesh_model_pub_is_unicast(dimmer->lvl_client.model)) {
+		bt_mesh_lvl_cli_move_set(&dimmer->lvl_client, NULL, &move_set, NULL);
+	} else {
+		bt_mesh_lvl_cli_move_set_unack(&dimmer->lvl_client, NULL, &move_set);
+	}
+}
+
+static struct dim_button buttons[] = {
+
+	/* Button 1 */
+	{ .lvl_client = BT_MESH_LVL_CLI_INIT(NULL),
+	  .onoff_client = BT_MESH_ONOFF_CLI_INIT(NULL),
+	  .press_start_time = 0,
+	  .direction = DIM_DOWN },
+
+	/* More dimmer buttons can be added here and may be used
+	 * simultaneously.
+	 */
```

Comment:
Using an array for this does not extend well to supporting buttons for other types of client models, which we want to do here in the future. Suggest to just define a single instance here and a simple `if (changed & BIT(0))` in the button handler, removing the loop there. We can add the array back easily enough if we need it in the future.